{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "accompanied-antarctica",
   "metadata": {},
   "source": [
    "## Description\n",
    "\n",
    " - The code is adapted from a task to find the most conserved regions of 100-200 bp in 10 SARS-Cov-2 genomes without performing MSA.\n",
    " - A total of eleven SARS-Cov-2 genomes were downloaded from NCBI [https://www.ncbi.nlm.nih.gov/sars-cov-2/] \n",
    " \n",
    "## Result & Discussion\n",
    "\n",
    " - A total of 184 conserved regions of 100 bp were mapped in the 10 genomes. Because the eD value is 0, these conserved regions are in fact identical in all genomes.\n",
    " - This result is expected due to the fact that the genomes we analyzed are simply potential different variants of the same virus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "supported-providence",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class genome(object):\n",
    "    \"\"\"read a .fasta and assign the sequence to a variable\"\"\"\n",
    "    def __init__(self, filename):\n",
    "        self.filename = filename\n",
    "        \n",
    "    def readgenome(self, filename):\n",
    "        gnome = ''\n",
    "        with open(filename, 'r') as f:\n",
    "            for line in f:\n",
    "                if not line[0] == '>':\n",
    "                    gnome += line.rstrip()\n",
    "        return gnome\n",
    "    \n",
    "\n",
    "def jjGlobal_editDist(x, y):\n",
    "    \"\"\"find the best alignment between x and y and return edit distance value and alignment position in y\"\"\"\n",
    "    D = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]* (len(y)+1)) \n",
    "        #create a matrix of all zeros that has (len(x)+1) rows and (len(y)+1) columns\n",
    "    \n",
    "    for i in range(1, len(x)+1):\n",
    "        D[i][0] = D[i-1][0] + 1 #first column with numerically increase numbers from (0, len(x)+1)\n",
    "    for i in range(1, len(y)+1):\n",
    "        D[0][i] = D[0][i-1] # the fist row remain zero values\n",
    "    \n",
    "    count = 0\n",
    "    for i in range(1, len(x)+1):\n",
    "        for j in range(1, len(y)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if x[i-1] == y[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else: \n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            \n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "            #count += 1\n",
    "        \n",
    "        D_pos = j - i\n",
    "            \n",
    "    return min(D[len(x)]), D_pos\n",
    "\n",
    "\n",
    "class g_eD(object):\n",
    "    \"\"\"generate a list for each genome (g) using the queries and Oleg_editDist function\"\"\"\n",
    "    def __init__(self, g):\n",
    "        self.g = g\n",
    "\n",
    "    def eD(self, g):\n",
    "        self.g = g\n",
    "        eD = []\n",
    "        for (seg, start) in ref_seg_id:\n",
    "            eD.append(Oleg_editDist(seg, g))\n",
    "        return eD\n",
    "    \n",
    "class eD_item (object):\n",
    "    \"\"\"for each eD, generate two lists containing the eD value and pos info, respectively\"\"\"\n",
    "    def __init__(self, eD):\n",
    "        self.eD = eD\n",
    "        \n",
    "    def eD_val(self, eD):\n",
    "        self.eD = eD\n",
    "        eD_val_list = []\n",
    "        eD_val_list.append([val for (val, pos) in eD])\n",
    "        return eD_val_list\n",
    "    \n",
    "    def eD_pos(self, eD):\n",
    "        self.eD = eD\n",
    "        eD_pos_list = []\n",
    "        eD_pos_list.append(pos for (val,pos) in eD)\n",
    "        return eD_pos_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "experimental-expert",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a list containing the 10 genome sequences: gnome_list\n",
    "\n",
    "g_name = ['MZ401444.fasta', 'MZ401459.fasta', 'MZ401492.fasta', 'MZ401506.fasta', 'MZ401850.fasta', 'MZ402172.fasta', 'MZ402197.fasta', 'MZ402259.fasta', 'MZ402465.fasta', 'MZ402490.fasta']\n",
    "gnome_list = []\n",
    "for name in g_name:\n",
    "    gn = genome(name)\n",
    "    gnome_list.append(gn.readgenome(gn.filename))\n",
    "    \n",
    "# unpack gnome_list so that each genome can be assigned to variables from g1 to g10 in the new list: g_list\n",
    "g1, g2, g3, g4, g5, g6, g7, g8, g9, g10 = gnome_list\n",
    "g_list = [g1, g2, g3, g4, g5, g6, g7, g8, g9, g10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "sonic-swiss",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain reference genome to be used for generation of queries\n",
    "\n",
    "ref = genome('NC045512_Ref.fasta')\n",
    "g_ref = ref.readgenome(ref.filename)\n",
    "\n",
    "def segment_id(genome, L):\n",
    "    \"\"\"generates a list of tuples (seq, pos) containing all segments (length=L) after slicing the genome \"\"\"\n",
    "    segment_id = []\n",
    "    sg_seq = ''\n",
    "    sg_start= ''\n",
    "    count = 0\n",
    "    for i in range(len(genome)//L):\n",
    "        sg_seq = genome[i*L:(i*L+L)]\n",
    "        count +=1\n",
    "        sg_start = i*L\n",
    "        segment_id.append((sg_seq,sg_start))  \n",
    "    \n",
    "    return segment_id\n",
    "\n",
    "ref_seg_id = segment_id(g_ref, 100) #list of query sequences with position info generated from the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "urban-eight",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a list of eD_list, and unpack so that eD1 = eD(g1), ...\n",
    "# eD1 is a list of tuples from g1, each tuple (minD, [D_loc]) corresponds to a query\n",
    "\n",
    "eD_list = []\n",
    "for g in g_list:\n",
    "    gEd = g_eD(g)\n",
    "    eD_list.append(gEd.eD(gEd.g))\n",
    "    \n",
    "eD1, eD2, eD3, eD4, eD5, eD6, eD7, eD8, eD9, eD10 = eD_list\n",
    "g_eD_list = [eD1, eD2, eD3, eD4, eD5, eD6, eD7, eD8, eD9, eD10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "italian-senator",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate the eD value and pos info lists for all genomes in two separate lists\n",
    "\n",
    "val_list = []\n",
    "pos_list = []\n",
    "for eD in g_eD_list:\n",
    "    gItem = eD_item(eD)\n",
    "    val_list.append(gItem.eD_val(gItem.eD))\n",
    "    pos_list.append(gItem.eD_pos(gItem.eD)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "republican-diploma",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine the lists of eD values and pos from the 10 genomes into arrays, respectively\n",
    "\n",
    "val_matrix = np.array(val_list, dtype=object)\n",
    "pos_matrix = np.array(pos_list, dtype=object)\n",
    "\n",
    "eD_sum = np.sum(val_matrix, axis=0) # return 1D array containing the sum of eD values for each query\n",
    "eD_min = np.amin(eD_sum) # find the lowest eD value in the array\n",
    "eD_min_index = np.where(eD_sum==eD_min) # Extract indices where the lowest eD value was observed for all genomes\n",
    "\n",
    "# 'eD_min_index' is a tuple containing two arrays, needs to extract the 2nd array, and converted it to a list\n",
    "(_,eD_index)= eD_min_index #extract the 2nd array in the tuple\n",
    "eD_index = eD_index.tolist() #change array to list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "iraqi-bedroom",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make df of the extracted array and use df._csv to save it as a .csv file (output)\n",
    "df = pd.DataFrame(con_region)\n",
    "df.to_csv('con_reg.csv') # the file is renamed to con_region.csv for submission\n",
    "\n",
    "# obtain the list of conserved sequences with numerical labels with was included in con_region.csv\n",
    "ref_seg = [seg for (seg, ID) in ref_seg_id]\n",
    "r_seg = np.array(ref_seg)\n",
    "c_seg = list(r_seg[eD_index])\n",
    "df = pd.DataFrame(c_seg)\n",
    "df.to_csv('con_regSeq.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
